L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorCoarsenerImpl.cc

```diff
@@ -0,0 +1,67 @@
+#include "L1Trigger/L1THGCal/interface/concentrator/HGCalConcentratorCoarsenerImpl.h"
+
+HGCalConcentratorCoarsenerImpl::
+HGCalConcentratorCoarsenerImpl(const edm::ParameterSet& conf)
+  : fixedDataSizePerHGCROC_(conf.getParameter<bool>("fixedDataSizePerHGCROC")),
+    coarseTCmapping_(std::vector<unsigned>{HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_,HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_,HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_,HGCalCoarseTriggerCellMapping::kCTCsizeVeryFine_})
+{    
+}
+
+
+void 
+HGCalConcentratorCoarsenerImpl::
+updateCoarseTriggerCellMaps(const l1t::HGCalTriggerCell& tc, uint32_t ctcid){
+  
+
+  coarseTCs_[ctcid].sumPt += tc.pt();
+  coarseTCs_[ctcid].sumHwPt += tc.hwPt();
+  coarseTCs_[ctcid].sumMipPt += tc.mipPt();
+
+  if ( tc.mipPt() > coarseTCs_[ctcid].maxMipPt ){
+    coarseTCs_[ctcid].maxId = tc.detId();
+    coarseTCs_[ctcid].maxHwPt = tc.hwPt();
+    coarseTCs_[ctcid].maxMipPt = tc.mipPt();
+  }
+
+}
+
+void
+HGCalConcentratorCoarsenerImpl::
+assignCoarseTriggerCellEnergy(l1t::HGCalTriggerCell &tc, uint32_t ctcid){
+    tc.setHwPt(coarseTCs_[ctcid].sumHwPt);
+    tc.setMipPt(coarseTCs_[ctcid].sumMipPt);
+    tc.setPt(coarseTCs_[ctcid].sumPt);
+}
+
+void 
+HGCalConcentratorCoarsenerImpl::
+select(const std::vector<l1t::HGCalTriggerCell>& trigCellVecInput, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput)
+{ 
+
+  // first pass, fill the coarse trigger cell information
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+    uint32_t ctcid = coarseTCmapping_.getCoarseTriggerCellId(tc.detId());
+    updateCoarseTriggerCellMaps( tc, ctcid );
+  }
+
+  for (const l1t::HGCalTriggerCell & tc : trigCellVecInput) {
```

Comment:
The reason is that we need to push back the actual TC. The alternative would be to create the trigger cell again, as is done in the SuperTriggerCell class. I am not sure what is more efficient.

Reply:
Anyway a new trigger cell is creating, either with copy constructor or an other constructor

Reply:
We have the issue that in certain modules we might not want to coarsen. Therefore we would need to keep the individual TC energies until this point. An example of what can nearly be done is here:

```
  for (auto & ctc : coarseTCs_){

      int thickness = triggerTools_.thicknessIndex(ctc.first.detId(),true);

      if ( fixedDataSizePerHGCROC_ && thickness == kHighDensityThickness_ ){

        std::vector<uint32_t> output_ids = coarseTCmapping_.getConstituentTriggerCells( ctc.first );
        for (const auto& id: output_ids){
          l1t::HGCalTriggerCell triggerCell;
          GlobalPoint point = triggerTools_.getTCPosition(id);
          math::PtEtaPhiMLorentzVector p4(0, point.eta(), point.phi(), 0.);
          triggerCell.setPosition(point);
          triggerCell.setP4(p4);
          triggerCell.setDetId(id);
          trigCellVecOutput.push_back( triggerCell );

          //NEED TO ASSIGN CORRECT ENERGIES
        }
        continue;
      }


      l1t::HGCalTriggerCell triggerCell;
      trigCellVecOutput.push_back( triggerCell );
      assignCoarseTriggerCellEnergy( trigCellVecOutput.back(), ctc.first );
      coarseTCmapping_.setEvenDetId(trigCellVecOutput.back());
      coarseTCmapping_.setCoarseTriggerCellPosition( trigCellVecOutput.back() );

  }
```

Reply:
I also wonder if certain properties of the TC would be lost, that are later needed somewhere (e.g. we could forget to set `mipPt` for example).

Reply:
In the proposed code TCs are created and their properties are then modified. It would be better to directly create the object with the correct properties (I had already the same comment somewhere else)

Reply:
I have now modified the other instance, so instead of `setCoarseTriggerCellPosition` I have `getCoarseTriggerCellPosition`, and use it as the GlobalPoint when creating the TC.